### ISSUE TITLE
ActivityPub replies

### ISSUE THREAD

**Issue by koehn** [friendica/friendica#8627]
### Expected behavior
Replies & reactions should be addressed to all recipients of the Document to which you are replying. 

### Actual behavior
Replies are not addressed this way, and seem to include Public and the replying actor's followers.

### Steps to reproduce the problem
Examine https://friendica.mrpetovan.com/objects/735a2029-395e-bb0d-cc65-a5e978200598
and its `inReplyTo`.

## Background
My little platform allows users to address messages to different audiences, similar to Diaspora and Facebook. Messages can be public, in which case they're available in the actor's outbox and addressed to the actor's followers, and they can also be restricted to specific actors. 

The message replied to above is public and addressed to my followers, but the reply is not. So while I can see Hypolite's reply, my other followers cannot, because they don't have the message, because it wasn't addressed to them. 

I choose to favor *graph consistency* in my approach: all Actors addressed in a "root" message (one not `inReplyTo` another message) _should_ receive the same graph of objects (Documents, Likes, etc.), so that they can be equal participants in the conversation, and have the opportunity to store those objects locally for fast retrieval/display. 

The algorithm for this approach is a simple one: you address responses to messages to the same audience as the message you're replying to, plus anyone you've tagged. 

Note that I am not saying that the sending server is responsible for *delivery* of all those messages: I am attempting to follow the ActivityPub guidance where if (for example) an Actor's followers are not visible to another server, that Actor's server is responsible for delivering the message to their followers. 

But by following this simple addressing algorithm, everyone is insured that non-public messages remain non-public (including their replies), and everyone can see all the relevant objects about a message. The graph of activityPub objects is (eventually) consistent among all participants in the conversation. It's roughly equivalent to "reply all" in email. 

I'd love to know your thoughts.

**Comment by MrPetovan** [comment]
Doesn't the author's followers appears as a meta-group in the object though? But the message was physically delivered to multiple actual inboxes. How is the replying system supposed to deliver the reply to the author's meta-group of followers? Do we have to store a local copy of all the relationships of all the accounts we know about? I do believe we are headed for this feature in Friendica but the sheer number of relationships is dizzying.

**Comment by annando** [comment]
I have the feeling as if at the beginning we had got the followers collection of the original author in our receivers as well - and if I remember correctly, people from the Mastodon and/or Pleroma project hadn't liked that.

**Comment by koehn** [comment]
I define the entries in `to`, `cc`, et all as "Addressees". An addressee can be an Actor, a Collection, or a special case like "public". I'm just following the ActivityPub specification for this. 

When _delivering_ an activity, I iterate through all the addressees and resolve them to actors, then look up the outbox for each of those actors and POST the activity there. If I come across a collection that I can resolve, I do so and send the message to every actor in the collection. The activity's fields the same for all recipients of the Activity.

This makes responding to an activity easy, since in most cases you simply copy the addressees of the activity/document to which you're replying and you're good to go. 

Otherwise, how should the receiving system determine if it should forward the message on, what should it put in the addressees? How does forwarding like that work if you're relying on http-signature for authentication/authorizaion?

**Comment by koehn** [comment]
Here's what I mean, specifically:

https://gitlab.koehn.com/mammoth/mammoth-server/-/blob/master/src/service/activity-transport.ts

**Comment by annando** [comment]
Forwarding for public posts always works, since the post URLs can be fetched. (What we do) For non public posts we do rely on LD signatures.

Especially on non public posts you cannot know all receivers. People could be added via `bcc`. Friendica (due to its history) is based on the inheritance of permissions from the thread starter post.

How do you resolve collections? Do you fetch them from the other accounts at the time of posting the message? if yes, then I see some problem there: At first, not all systems do publish data in these endpoints. And second: This list can be incomplete. Also this is no option for Friendica. Good systems can deliver around 1,000 messages per minute. Doing the same would surely dramatically reduce the performance. And we also couldn't simply cache and update the collection every few days, since the data could be outdated.

P.S.: Concerning the discussion: I have got a short attention span and I also have got problems understanding theoretical descriptions (I guess at some point in the past I nearly drove @MrPetovan and @nupplaphil crazy with my way of understanding). I mostly understand through examples.

**Comment by koehn** [comment]
> Especially on non public posts you cannot know all receivers. People could be added via bcc. Friendica (due to its history) is based on the inheritance of permissions from the thread starter post.

Yes, I assume that when a response is received that the originating system is responsible for additional delivery to e.g., bcc and bto recipients. 

> How do you resolve collections? 
I attempt to resolve them at delivery time, and if received, deliver the messages to the actors in the collection. If I the collection isn't resolvable, I assume it's the original sender's responsibility to do so. 

> Good systems can deliver around 1,000 messages per minute. Doing the same would surely dramatically reduce the performance. 
Somebody needs to deliver the messages, and it's every bit as likely that a Friendica user will have thousands of followers as anybody else, so I don't see what difference it makes if it's the message sender who expands the collection or the original Actor who does.  

> And we also couldn't simply cache and update the collection every few days, since the data could be outdated.
Agreed. I don't cache collections, though I do cache items in collections for about an hour. 

As an example, I point to the above message: I sent it to 'public' and my followers collection. The `to` looks like: 
```
  "to": [
    "https://www.w3.org/ns/activitystreams#Public",
    "https://mammoth.home.koehn.com/api/actor/bkoehn/followers"
  ],
```
I have two followers, Hypolite and Alice, both of whom received the message. Hypolite responded using Friendica, and his `to` looked like:
```
  "to": [
    "https://mammoth.home.koehn.com/api/actor/bkoehn",
    "https://www.w3.org/ns/activitystreams#Public",
    "https://friendica.mrpetovan.com/followers/hypolite"
  ],
```
(the `cc` was an empty array)

So the message wasn't delivered to Alice. As far as she can tell, nobody replied to it. If she replied to the message, her server would address the message to me, 'public' and my followers, and Hypolite would have gotten the her reply. At this point, Hypolite and I would have complete graphs of the messages and replies; Alice wouldn't, but she has no way of knowing that she doesn't. 

Hypolite and I both replied to his reply, my server copying the recipients from the previous message, meaning that Alice is dropped off completely. My server had no idea that it even *should* forward the message to Alice, since how would it distinguish between the case where it should and the case where it shouldn't? Had Friendica included my followers collection on the `to` address, it still would have assumed that Friendica delivered the message to the members of that collection, because why wouldn't it? Friendica is perfectly capable of doing that. Had my Actor restricted access to the followers collection, and it saw a message including my Actor's follower collection as a recipient, it would have done the delivery itself. Otherwise it has no way of knowing that it has any responsibility to deliver the message to additional recipients at all.

**Comment by koehn** [comment]
I can imagine a middle ground, where the sender indicates that they aren't delivering the messages to all the recipients could be used; then the system that sent the original message would know that it's expected to deliver the message, e.g., 
```
to: [
    "recipientsOf:https://mammoth.home.koehn.com/api/actor/bkoehn/79ae23ff-0fc4-4f81-b2be-038562bae23c"
]
```

That way it's possible to keep a unified addressee. You could POST that message to my user's `inbox` and it would know that it's expected to take it from here, and what to do. We'd have to figure out some way to prevent spam tho, and it unnecessarily slows down the delivery process by adding a hop. 

Adding a `ewply-to` field on Objects would make a lot of sense as well.

**Comment by annando** [comment]
Concerning the resolving of collections: In the past a server owner put my server on the blocklist, because I had resolved the collections. He said that the privacy terms on his instance do forbid fetching data from his instance. The situation had been a little bit different, since my script went crazy and constantly polled the data (which he then noticed - which caused his investigations).

I see a huge problem with resolving the collections. Say you deliver a post to someone with 1,000 followers. And that system has got a paging after 50 entries. This means that you have to poll the collection 20 times to get the full list. This would reduce the delivery times a lot. And then imagine that all systems would do the same. This would but some systems under real stress.

I completely do understand your point. But I also see problems with the concept.

BTW: expanding the "to" to personal inboxes will create some problem with Pleroma (AFAIK). Pleroma defines the "to" equivalent to tagging an account. This means that these posts would appear like a mention, although that person possibly wasn't mentioned at all.

In the past there had been this rule:
* to: All persons that are mentioned and that should be notified about this post. When the public collection is in "to", then this post is completely public and will appear on the public timelines other servers.
* cc: The rest of the (visible) receivers. If the public collection is in "cc" then this post will still be public but won't appear on the public collection of other systems. In Friendica this alo means that these posts aren't visible for the tag subscription.

**Comment by koehn** [comment]
> I see a huge problem with resolving the collections. Say you deliver a post to someone with 1,000 followers. And that system has got a paging after 50 entries. This means that you have to poll the collection 20 times to get the full list. This would reduce the delivery times a lot. And then imagine that all systems would do the same. This would but some systems under real stress.

Are you suggesting that only the originating system resolve the collection and put 1000 actors in `to` or `cc`? That would massively increase the message payload, especially as 1000 followers becomes 50,000. A simple 'Like' message with 50,000 names in the `cc` field would be 2.5MB, per like. There's simply no way to scale that solution, and almost all of the information in each Like is unnecessary. That's why I never even considered it as a viable option. 

You'll save a ton of disk space and bandwidth simply including the name of the collection instead, and having an efficient API for resolving it, once per message. 

> BTW: expanding the "to" to personal inboxes will create some problem with Pleroma (AFAIK). Pleroma defines the "to" equivalent to tagging an account. This means that these posts would appear like a mention, although that person possibly wasn't mentioned at all.

That's a choice. Thanks for the heads-up; I'm not planning on expanding the collection at all tho. The collection is a valid addressee; it should be preserved and supported. (Hence this issue).

My implementation follows my understanding of the spec: if the "public" recipient is in `to` or `cc`, the item is in the outbox, and eligible for being pulled, indexed, etc. Mentions composed by users may be written as `@user@example.com` and will result in an entry in the `to` field and an entry in the `tag` field as well. 

In the large, it seems like every implementation of ActivityPub bent the spec to solve certain implementation challenges. I'm certain my implementation will add its own to the mix; time will tell how they resolve.

**Comment by MrPetovan** [comment]
I believe what @annando is saying that your system should "forward" my reply to its known list of recipients using local relationship data. This would probably mean tacking `"https://mammoth.home.koehn.com/api/actor/bkoehn/followers"` in the `cc` field and would be more efficient since there wouldn't be repeated API hits to fetch the complete list of followers from your system.

I believe the problem of fetching tagged recipients' followers would also become exponentially worse with the amount of people involved in a thread. Letting each originating system forward the message to the followers of their local users would reduce (no followers API hits) and distribute the load.

**Comment by koehn** [comment]
> I believe what @annando is saying that your system should "forward" my reply to its known list of recipients using local relationship data. This would probably mean tacking "https://mammoth.home.koehn.com/api/actor/bkoehn/followers" in the cc field and would be more efficient since there wouldn't be repeated API hits to fetch the complete list of followers from your system.

How should my system know that your system expects my system  to do this, and that the message wasn't meant to be handled as delivered? It's the unstated-ness of this method that I find objectionable, since it requires every future developer to learn an undocumented set of rules. 

> I believe the problem of fetching tagged recipients' followers would also become exponentially worse with the amount of people involved in a thread. 

Why should the system send a message to tagged recipients followers? They're following the recipient, not the sender. 

> Letting each originating system forward the message to the followers of their local users would reduce (no followers API hits) and distribute the load.

1. There's no reason to believe that followers are local
2. Even if some are, there's no reason to believe that this would substantially reduce the load. 
3. If they aren't local, the load on the originating server goes up exponentially with the number of active followers
4. There's no reason to believe API hits to gather followers is somehow more expensive than making the sending server manage deliveries all by itself

Let's take followers out of the discussion for a moment, and talk about audience. Do I, as a meessage sender, deserve to know who the audience is for my message? I can learn many recipients based simply on the reactions, but shouldn't I be allowed to know who is going to see my message before I send it? If the answer to that question is "yes", then I'd like to propose a different solution:

What if we allowed systems to access the recipient list of any message they're allowed to reply to? With the exception of `bcc` and `bto` recipients, if I'm able to do a `GET` query on a message where I'm a recipient, I should be able to get a Collection of recipients (expanded to Actors, not impenetrable Collections) from the originating system. Using that Collection my server can deliver the message on its own. 

So for a message with id `https://example.com/objects/12345` might look like:
```
  "id": "https://example.com/objects/12345",
  "type": "Note",
  "content": "Hello, world!",
  "to": "https://example.com/objects/12345/recipients",
  "reply-to": "https://example.com/objects/12345/recipients"
}
```
Then servers (and clients) could `GET https://example.com/objects/12345/recipients` to get the complete Collection/PagedCollection of Actors (in this case, all the actors in the `https://example.com/actors/brad/followers` collection). Servers could do an authorization check to make sure the actor is authorized to view the recipients. From there, not only can the user determine who they're talking to, the server can deliver messages directly. 

Servers *could* choose to allow the `reply-to` to accept `POST` requests, and deliver messages to those addresseees for the other server. When disallowed (with a `Forbidden` response), replying servers must expand the reply-to collection themselves and deliver the message to the actors in that collection. 

This proposal provides a hopefully reasonable middle ground, where Followers collections may be kept private, users know with whom they're communicating, and server owners can choose to either distribute messages on behalf of another server or not. 

Thoughts on that?

**Comment by MrPetovan** [comment]
I don't have a sufficiently detailed knowledge of how AP works or "should" work or even does work in Friendica so I'm a little out of my depth but I'll try to provide meaningful answers.

> How should my system know that your system expects my system to do this, and that the message wasn't meant to be handled as delivered? It's the unstated-ness of this method that I find objectionable, since it requires every future developer to learn an undocumented set of rules.

I understand, and I don't know how we decide to forward replies to the original thread starter's followers or if we even actually do it, so I cannot comment further on that.

I do like your proposal that allows both actor fetching and forward. Could there even be a `Forbidden` response on the `GET /recipients` call for private mailing list with BCC for example, which would force the receiving server to `POST` the reply back to the originating server?

However I'm not sure about threads with more than 2 people.

Let's say Alice starts a thread and sends it to her followers. At this point `/recipients` contains only Alice's followers. Bob replies to it and send it to both Alice, her followers and his followers. Shouldn't the `/recipients` feature Bob's followers as well? If yes, how Alice's server would know?

**Comment by koehn** [comment]
> Let's say Alice starts a thread and sends it to her followers. At this point /recipients contains only Alice's followers. Bob replies to it and send it to both Alice, her followers and his followers. Shouldn't the /recipients feature Bob's followers as well? If yes, how Alice's server would know?

Great example! The recipients list is not *per thread*, but *per message*. Bob is not able to modify the audience of Alice's message, but may set the audience of his reply. If the message from Alice was not public, they might have no idea what Bob's message means, since they may not be able to access Alice's message. 

As a courtesy, he *should* maintain the set of recipients from Alice's message, so that they can maintain the same graph of messages. This is similar to email (and the way AP addresses messages), where users can "reply all" or just "reply" if they want. I would hope that most of the time they would "reply all" because that enables everyone to contribute to the conversation. 

In any case, recipients of Bob's message can query its recipients list, and reply to it.

**Comment by MrPetovan** [comment]
Ah, per message, fair enough. So that would be another n-m relationship to store locally and I don't know if I'm aroused at the normalization prospect or horrified at the potential disk space footprint of such a feature.

**Comment by koehn** [comment]
You have that problem today. Friendica already replies to my message with a different set of recipients than the ones on the message it's replying to. That's the whole point of my issue.

**Comment by koehn** [comment]
In any case, a "Post" is a directed, (hopefully) acyclic graph assembled from Activities and Objects.  At the root is the main post (which has no `inReplyTo`), and other Activites "attach" to it through either the `inReplyTo` field (for replies like `Note`s) or the `object` field (for `Like`s). Each of these can, in turn, have other objects attached, in a composite pattern. ActivityPub allows for fairly ridiculous constructs (it's possible to reply to a `Like`), so implementations have to choose what parts of the graph they're willing to display. 

In my little app, I store the Objects nearly the same way as they arrive: I unpack `Create` and `Update` objects separately from their container, but everything else gets persisted as received. I have a handful of views that quickly assembles everything into the correct DAG on demand. 

Addressing, however, is a different kettle of fish. AP requires each object be addressed individually, and provides no guidance for how messages should be addressed, nor any requirement for keeping a DAG consistent for different users. It's all up to us to decide, and so far we've all chosen very inconsistently, with little thought to the consequences.

**Comment by MrPetovan** [comment]
I understand. It's even more complex for a system like Friendica that had its own very flawed message persistance scheme before implementing AP, so we're trying to map AP concepts onto our existing model. It requires both an intimate knowledge of the Friendica model and AP concepts that so far only @annando has been able to meaningfully work on it.

**Comment by koehn** [comment]
Yeah, I'm not going to implement multi-protocol because you end up coming face-to-face with all the impedence mismatches between these systems. Just AP is hard enough.
